### ISSUE TITLE
Respect replaceOnChanges when calculating dependent replacements of DBRs

### ISSUE THREAD

**Issue by flostadler** [pulumi/pulumi#16764]
### What happened?

Right now replaceOnChanges isn't taken into account when traversing the dependencies of a resource that will be replaced with delete before replace (DBR).
This will lead to wrong replace orders and failures if the cloud providers verify whether a resource is still in use.
The bridge turns on DBR for manually named resources (see https://github.com/pulumi/pulumi-terraform-bridge/commit/96182a46b78c908f82f4b16af991aae8a60311d3) amplifying this issue for bridged providers.

Let's take this pulumi AWS program for example:
```
const tg = new aws.alb.TargetGroup("my-test-tg", {
    name: "my-test-tg"
    port: 443,
    protocol: "HTTP",
    vpcId: "VPC_ID",
    targetType: "ip",
});

const listener = new aws.alb.Listener("http", {
    loadBalancerArn: alb.arn,
    port: 80,
    protocol: "HTTP",
    defaultActions: [{
        type: "forward",
        targetGroupArn: tg.arn,
    }],
}, { replaceOnChanges: ["defaultActions[*].targetGroupArn"] });
```

The Target Group is marked as delete before replace because it has a fixed name. When updating properties that cause replacements (e.g. port) Pulumi should first delete the listener (because it is using the target group) and then delete the target group itself before it re-creates them in the opposite order.

Right now it's trying to delete the Target Group first which fails because it's still used by the listener. We're not correctly detecting that the listener needs to be replaced in tandem with the target group because replaceOnChanges isn't taken into account when evaluating the dependent replaces of a DBR.

### Example

```
const tg = new aws.alb.TargetGroup("my-test-tg", {
    name: "my-test-tg"
    port: 443,
    protocol: "HTTP",
    vpcId: "VPC_ID",
    targetType: "ip",
});

const listener = new aws.alb.Listener("http", {
    loadBalancerArn: alb.arn,
    port: 80,
    protocol: "HTTP",
    defaultActions: [{
        type: "forward",
        targetGroupArn: tg.arn,
    }],
}, { replaceOnChanges: ["defaultActions[*].targetGroupArn"] });
```

### Output of `pulumi about`

n/a

### Additional context

Causes:
- https://github.com/pulumi/pulumi/issues/16667
- https://github.com/pulumi/pulumi-aws/issues/4220
- https://github.com/pulumi/pulumi/issues/11273
- https://github.com/pulumi/pulumi/issues/14050
- https://github.com/pulumi/pulumi/issues/13972

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by Frassle** [comment]
I don't think this is doable right now. ReplaceOnChanges is only sent in the resource request, it's not saved as part of the state. When the TargetGroup resource registration comes in the engine can't have seen the Listener registration yet (because the program can't send it till the TargetGroup resolves). 

If we saved ReplaceOnChanges to state, we could use the saved state to help decided replacement chains.
